tests/FunctionalTestCase.php

```diff
@@ -513,6 +533,26 @@ protected function skipIfTransactionsAreNotSupported(): void
         }
     }
 
+    private static function appendAuthenticationOptions(array $options): array
+    {
+        if (isset($options['username']) || isset($options['password'])) {
+            return $options;
```

Comment:
Perhaps it makes sense to raise an exception here, or maybe only if the options are later specified and would try to append themselves?

I reckon there isn't a use case where we'd actually want `MONGODB_USERNAME` or `MONGODB_PASSWORD` to override anything, apart from the previous edge case I mentioned about the connection string already having auth credentials.

Reply:
This method is called automatically, and only appends default options if the URI being given doesn't already contain them. This allows us to specifically test with invalid credentials on clusters with auth enabled, which would otherwise be impossible (either due to the exception we'd raise, or because we'd always inject the credentials from the environment).

Reply:
Sorry, I should've clarified. Yes, this does not catch credentials sent through the connection string, which we currently don't do in tests. This ensures that any client created with URI options that set specific credentials (e.g. to test failing authentication) does not get correct credentials injected.